src/components/implementation/fork_tests/voter/voter.c

```diff
@@ -0,0 +1,290 @@
+		if (c->have_A2B_data) {
+			c->have_A2B_data = 0;
+			return c->A2B_data;
+		}
+		else {
+			ret = block_replica(replica);
+			if (ret < 0) printc("Error");
+			/* On wakeup */	
+			assert(c->have_A2B_data);
+			c->have_A2B_data = 0;
+			return c->A2B_data;
+		}
+	}
+	else {
+		BUG();
+		return -1; // shut up compiler
+	}
+}
+
+int confirm(spdid_t spdid, replica_type type) {
+	printc("Confirming readiness of spdid %d\n", spdid);
+	
+	int ret;
+	int i;
+	
+	int t = (type == ping) ? 0 : 1;	
+	struct replica_info *replicas = components[t].replicas;
+	
+	for (i = 0; i < components[t].nreplicas; i++) {
+		if (replicas[i].spdid == 0) {
+			printc("creating replica for spdid %d in slot %d\n", spdid, i);
+			replicas[i].spdid = spdid;
+			//replicas[i].spdid = quarantine_fork(cos_spd_id(), comp2fork);
+			//if (replicas[i].spdid == 0) printc("Error: f1 fork failed\n");
+			replicas[i].buf_read = cbuf_alloc(1024, &replicas[i].read_buffer); 
+			replicas[i].buf_write = cbuf_alloc(1024, &replicas[i].write_buffer);
+			replicas[i].blocked = 0;
+			replicas[i].thread_id = cos_get_thd_id();
+
+			map[spdid] = type;
+			return 0;
+		}
+		else {
+			printc("replica for spdid %d already exists\n", replicas[i].spdid);
+		}
+	}
+	return 0;
+}
+
+void monitor(void) {
+	struct nmod_comp *rcv_comp;
+	struct replica_info *rcv_replica;
+	int i, j;
+	int found = 0;
+
+	for (i = 0; i < n_channels; i++) {
+		printc("Examing channel %d between nmodcomps with replicas %d and %d \n", i, channels[i].A->replicas[0].spdid, channels[i].B->replicas[0].spdid);
+
+		printc("channel A2B_data present %d (%d)\n", channels[i].have_A2B_data, channels[i].A2B_data);
+		printc("channel B2A_data present %d (%d)\n", channels[i].have_B2A_data, channels[i].B2A_data);
+
+		if (channels[i].have_A2B_data) {
+			rcv_comp = channels[i].B;
+			found = 1;
+		}
+		else if (channels[i].have_B2A_data) {
+			rcv_comp = channels[i].A;
+			found = 1;
+		}
+
+		if (found) {
+			printc("Examining receiving comp with replicas %d\n", rcv_comp->nreplicas);
+
+			for (j = 0; j < rcv_comp->nreplicas; j++) {
+				rcv_replica = &rcv_comp->replicas[j];
+
+				printc("Looking at replica %d\n", rcv_replica->spdid);
+
+				if (!rcv_replica->thread_id) BUG();
+
+				if (rcv_replica->blocked) {
+					rcv_replica->blocked = 0;
+					printc("waking up thread %d\n", rcv_replica->thread_id);
+					sched_wakeup(cos_spd_id(), rcv_replica->thread_id);
+				}
+			}
+		}
+	}
+}
+
+void cos_init(void)
+{
+	int i, j;
+	for (i = 0; i < n_comps; i++) {
+		printc("Setting up nmod component %d\n", i);
+		components[i].nreplicas = 1; // for now, no fork
+		components[i].replicas = (struct replica_info*) malloc(sizeof(struct replica_info) * components[i].nreplicas);
+	
+		for (j = 0; j < components[i].nreplicas; j++) {
+			components[i].replicas[j].spdid = 0;
+			components[i].replicas[j].blocked = 0;
+		}
+	}
+
+	channels[0].A = &components[0];
+	channels[0].B = &components[1];
+
+	for (i = 0; i < 250; i++) map[i] = none;
```

Comment:
For the record, it is defined in interfaces/voter/voter.h but it is a terrible choice of name. I want to change this whole thing because interfaces shouldn't have information about ping/pong/none. This should likely be in some separate config file.